### ISSUE TITLE
panic: runtime error: slice bounds out of range

### ISSUE THREAD

**Issue by pouladzade** [hyperledger/burrow#697]
Please include in your bug report:

- [0.17.1 ] `burrow version` (docker image tag or branch if built from source)
- [ No] `monax version` (if applicable)
- [1.9.4 ] `go version` (if applicable)
- [No ] `docker version` (if applicable)
- [Ubuntu 16.4 LTS amd64 ] operating system details (osx/windows/linux)
- [ ] what you expected to happen
- [ ] what actually happened
- [manipulating bytes in solidity assembly ] steps to reproduction

[burrow-log.txt](https://github.com/hyperledger/burrow/files/1759074/burrow-log.txt)

```
DBUG[02-26|21:48:09] Account has permission                   log_channel=Trace run_id=aaa20d58-1afb-11e8-801c-d770f71e1cc1 trace="[convention.go:35 log.go:134 convention.go:50 convention.go:35 execution.go:1019 execution.go:1048 execution.go:400 burrow-mint.go:120 local_client.go:74 app_conn.go:73 execution.go:93 execution.go:29 execution.go:212 replay.go:360 replay.go:346 replay.go:296 replay.go:235 multi_app_conn.go:99 service.go:99 node.go:84 tendermint.go:119 consensus.go:37 core.go:63 serve.go:159 serve.go:177 command.go:635 command.go:710 command.go:669 burrow.go:45 main.go:22 proc.go:195 asm_amd64.s:2337]" scope="[Serve Core BurrowMintPipe BurrowMint ExecTx]" account_address=C43A1F81D8AB6C75B7A1916A95BED38A5EA5FFA8 perm_flag=call
DBUG[02-26|21:48:09]                                          log_channel=Trace run_id=aaa20d58-1afb-11e8-801c-d770f71e1cc1 trace="[execution.go:445 burrow-mint.go:120 local_client.go:74 app_conn.go:73 execution.go:93 execution.go:29 execution.go:212 replay.go:360 replay.go:346 replay.go:296 replay.go:235 multi_app_conn.go:99 service.go:99 node.go:84 tendermint.go:119 consensus.go:37 core.go:63 serve.go:159 serve.go:177 command.go:635 command.go:710 command.go:669 burrow.go:45 main.go:22 proc.go:195 asm_amd64.s:2337]" scope="[Serve Core BurrowMintPipe BurrowMint ExecTx]" output_account="Account{269A5DA1325AFF9D3D1E1429577EFF5C9F0AE17A:<nil> B:0 C:1241 S: P:{Base: 11111111111111; Set: 11111111111111 []}}"
DBUG[02-26|21:48:09] Calling existing contract                log_channel=Trace run_id=aaa20d58-1afb-11e8-801c-d770f71e1cc1 trace="[convention.go:35 log.go:134 convention.go:50 convention.go:35 execution.go:505 burrow-mint.go:120 local_client.go:74 app_conn.go:73 execution.go:93 execution.go:29 execution.go:212 replay.go:360 replay.go:346 replay.go:296 replay.go:235 multi_app_conn.go:99 service.go:99 node.go:84 tendermint.go:119 consensus.go:37 core.go:63 serve.go:159 serve.go:177 command.go:635 command.go:710 command.go:669 burrow.go:45 main.go:22 proc.go:195 asm_amd64.s:2337]" scope="[Serve Core BurrowMintPipe BurrowMint ExecTx]" contract_address=000000000000000000000000269A5DA1325AFF9D3D1E1429577EFF5C9F0AE17A contract_code=606060405260043610610062576000357C0100000000000000000000000000000000000000000000000000000000900463FFFFFFFF16806324A04ECE1461006757806377AAE9C914610096578063917A781F14610148578063F9FBD554146101F1575B600080FD5B341561007257600080FD5B61007A610333565B604051808260FF1660FF16815260200191505060405180910390F35B34156100A157600080FD5B610146600480803590602001909190803590602001908201803590602001908080601F0160208091040260200160405190810160405280939291908181526020018383808284378201915050505050509190803590602001909190803590602001908201803590602001908080601F0160208091040260200160405190810160405280939291908181526020018383808284378201915050505050509190505061033C565B005B341561015357600080FD5B6101EF600480803590602001909190803590602001908201803590602001908080601F0160208091040260200160405190810160405280939291908181526020018383808284378201915050505050509190803590602001908201803590602001908080601F0160208091040260200160405190810160405280939291908181526020018383808284378201915050505050509190505061039E565B005B34156101FC57600080FD5B61024C600480803590602001908201803590602001908080601F01602080910402602001604051908101604052809392919081815260200183838082843782019150505050505091905050610404565B604051808060200180602001838103835285818151815260200191508051906020019080838360005B83811015610290578082015181840152602081019050610275565B50505050905090810190601F1680156102BD5780820380516001836020036101000A031916815260200191505B50838103825284818151815260200191508051906020019080838360005B838110156102F65780820151818401526020810190506102DB565B50505050905090810190601F1680156103235780820380516001836020036101000A031916815260200191505B5094505050505060405180910390F35B60006008905090565B600060208381151561034A57FE5B049050600060208481151561035B57FE5B06111561036B5780806001019150505B60018383525B85850151602082028401526020860395506001820391506001810190506000821161037157505050505050565B6000602083518115156103AD57FE5B0490506000602084518115156103BF57FE5B0611156103CF5780806001019150505B60006001820191505B6020826020020384019050805185840152602085039450600182039150600182116103D8575050505050565B61040C610485565B610414610499565B600060806040518059106104255750595B9080825280601F01601F19166020018201604052509250608060405180591061044B5750595B9080825280601F01601F19166020018201604052509150835190506104726080858561039E565B61047F608084838561033C565B50915091565B602060405190810160405280600081525090565B6020604051908101604052806000815250905600A165627A7A723058200160F96CC2CE68D29187FB64975569B0C810EC93F8480B3FC8B44DD0A2282ECD0029
DBUG[02-26|21:48:09]                                          log_channel=Trace run_id=aaa20d58-1afb-11e8-801c-d770f71e1cc1 trace="[execution.go:510 burrow-mint.go:120 local_client.go:74 app_conn.go:73 execution.go:93 execution.go:29 execution.go:212 replay.go:360 replay.go:346 replay.go:296 replay.go:235 multi_app_conn.go:99 service.go:99 node.go:84 tendermint.go:119 consensus.go:37 core.go:63 serve.go:159 serve.go:177 command.go:635 command.go:710 command.go:669 burrow.go:45 main.go:22 proc.go:195 asm_amd64.s:2337]" scope="[Serve Core BurrowMintPipe BurrowMint ExecTx]" callee_=(MISSING)
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
github.com/hyperledger/burrow/manager/burrow-mint/evm.(*dynamicMemory).Read(0xc421e44b00, 0xffffffffffffffe0, 0x20, 0xc421e60e30, 0x3, 0x3, 0x0, 0x0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/manager/burrow-mint/evm/memory.go:67 +0x178
github.com/hyperledger/burrow/manager/burrow-mint/evm.(*VM).call(0xc421e62240, 0xc4216e9480, 0xc4216e9500, 0xc42177c500, 0x4d9, 0x4d9, 0xc42001d040, 0x44, 0x44, 0x0, ...)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/manager/burrow-mint/evm/vm.go:657 +0x13520
github.com/hyperledger/burrow/manager/burrow-mint/evm.(*VM).Call(0xc421e62240, 0xc4216e9480, 0xc4216e9500, 0xc42177c500, 0x4d9, 0x4d9, 0xc42001d040, 0x44, 0x44, 0x0, ...)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/manager/burrow-mint/evm/vm.go:155 +0x218
github.com/hyperledger/burrow/manager/burrow-mint/state.ExecTx(0xc4201ffbf0, 0x11aaf00, 0xc421c36e60, 0xc4213b1901, 0x11ac180, 0xc4201ffe00, 0x11b7f00, 0xc421ad0a80, 0xc4213b53f0, 0xc421e96ca0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/manager/burrow-mint/state/execution.go:521 +0x3535
github.com/hyperledger/burrow/manager/burrow-mint.(*BurrowMint).DeliverTx(0xc420083770, 0xc4201741e0, 0xf0, 0xf0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/manager/burrow-mint/burrow-mint.go:120 +0x2b2
github.com/hyperledger/burrow/vendor/github.com/tendermint/abci/client.(*localClient).DeliverTxAsync(0xc4204deb40, 0xc4201741e0, 0xf0, 0xf0, 0x0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/abci/client/local_client.go:74 +0x85
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/proxy.(*appConnConsensus).DeliverTxAsync(0xc420049500, 0xc4201741e0, 0xf0, 0xf0, 0xc42133db00)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/proxy/app_conn.go:73 +0x51
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/state.execBlockOnProxyApp(0x0, 0x0, 0x11b8980, 0xc420049500, 0xc421e85d40, 0xc421ad0780, 0x11aa300, 0xc4214cd710)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/state/execution.go:93 +0x749
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/state.(*State).ValExecBlock(0xc4204dd2b0, 0x0, 0x0, 0x11b8980, 0xc420049500, 0xc421e85d40, 0xc420170e20, 0xc421e85d40, 0x0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/state/execution.go:29 +0x82
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/state.(*State).ApplyBlock(0xc4204dd2b0, 0x0, 0x0, 0x11b8980, 0xc420049500, 0xc421e85d40, 0x1, 0xc421e96960, 0x14, 0x14, ...)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/state/execution.go:212 +0x73
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/consensus.(*Handshaker).replayBlock(0xc4201fed80, 0x99c, 0x11b8980, 0xc420049500, 0x14, 0x20, 0x0, 0x0, 0xc420280a88)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/consensus/replay.go:360 +0x108
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/consensus.(*Handshaker).replayBlocks(0xc4201fed80, 0x11ba2c0, 0xc4204ae000, 0x0, 0x99c, 0x1, 0x0, 0x0, 0x4, 0x4, ...)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/consensus/replay.go:346 +0x2d8
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(0xc4201fed80, 0x0, 0x0, 0x0, 0x0, 0x11ba2c0, 0xc4204ae000, 0x0, 0x0, 0x0, ...)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/consensus/replay.go:296 +0x67a
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0xc4201fed80, 0x11ba2c0, 0xc4204ae000, 0x0, 0x0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/consensus/replay.go:235 +0x41a
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/proxy.(*multiAppConn).OnStart(0xc4204ae000, 0xc420155f60, 0x15)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/proxy/multi_app_conn.go:99 +0x1eb
github.com/hyperledger/burrow/vendor/github.com/tendermint/go-common.(*BaseService).Start(0xc4204ae000, 0x0, 0x0, 0x0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/go-common/service.go:99 +0x290
github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/node.NewNode(0x11bac20, 0xc4204cc2b8, 0xc4200a2820, 0x11ac380, 0xc4204f33c0, 0x2)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/tendermint/tendermint/node/node.go:84 +0x58e
github.com/hyperledger/burrow/consensus/tendermint.NewTendermint(0xc4204d3700, 0x11b9bc0, 0xc420083770, 0x11b7f00, 0xc42021e780, 0xb34260, 0xe19e70, 0xb34260)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/consensus/tendermint/tendermint.go:119 +0xb4d
github.com/hyperledger/burrow/consensus.LoadConsensusEngineInPipe(0xc4204d3700, 0x11bacc0, 0xc4204d2600, 0x24, 0x0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/consensus/consensus.go:37 +0x180
github.com/hyperledger/burrow/core.NewCore(0xc4201e0790, 0xa, 0xc4204d3700, 0xc4204d3780, 0x11b7f00, 0xc42021f340, 0x4, 0x0, 0x0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/core/core.go:63 +0x2fe
github.com/hyperledger/burrow/cmd.NewCoreFromDo(0xc4204ded80, 0x7ffe9ac2204e, 0x13, 0xc54bfa)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/cmd/serve.go:159 +0x9a1
github.com/hyperledger/burrow/cmd.ServeRunner.func1(0xc4201b66c0, 0xc4204d7f20, 0x0, 0x2)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/cmd/serve.go:177 +0xaf
github.com/hyperledger/burrow/vendor/github.com/spf13/cobra.(*Command).execute(0xc4201b66c0, 0xc4204d7ea0, 0x2, 0x2, 0xc4201b66c0, 0xc4204d7ea0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/spf13/cobra/command.go:635 +0x23d
github.com/hyperledger/burrow/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x11fe940, 0x1, 0x1, 0xc4201b66c0)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/spf13/cobra/command.go:710 +0x2fe
github.com/hyperledger/burrow/vendor/github.com/spf13/cobra.(*Command).Execute(0x11fe940, 0x0, 0xc4204e7f20)
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/vendor/github.com/spf13/cobra/command.go:669 +0x2b
github.com/hyperledger/burrow/cmd.Execute()
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/cmd/burrow.go:45 +0x312
main.main()
	/home/ahmad/projects/golang/src/github.com/hyperledger/burrow/cmd/burrow/main.go:22 +0x20
```

Contract that caused the issue, calling `test`:

```solidity
pragma solidity ^0.4.16;

contract StringTest{
    
    function test(string _in ) public returns( bytes memory _out,string memory _str){
        
        _out = new bytes(128);
        _str = new string(128);
        uint _size= bytes(_in).length;
        stringToBytes(128,bytes(_in),_out);
        bytesToString(128,_out,_size,bytes(_str));
        
    }
    
    function stringToBytes(uint256 _offst, bytes memory _input, bytes memory _output) public  {
        uint256 stack_size = _input.length / 32;
        if(_input.length % 32 > 0) stack_size++;
        
        assembly {
            let chunk := 0
            stack_size := add(stack_size,1)//adding because of 32 first bytes as the length
        loop:
            chunk := add(_input,sub(mul(32 ,stack_size),32))
            mstore(add(_output, _offst), mload(chunk))
            _offst := sub(_offst , 32)
            stack_size := sub(stack_size ,1)
            jumpi(loop , gt(stack_size,1))
        }
    }
    
    function bytesToString(uint256 _offst, bytes memory _input,uint256 _size, bytes memory _output) public  {

        uint256 stack_size = _size/32;
        if(_size % 32 > 0) stack_size++;
        
        assembly {
            let chunk_index := 1
            mstore(_output, _size)
            
        loop:
           mstore(add(_output,mul(chunk_index,32)), mload(add(_input,_offst)))
           _offst := sub(_offst , 32)
           stack_size := sub(stack_size ,1)
           chunk_index := add(chunk_index,1)
           jumpi(loop , gt(stack_size,0))
        }
    }
    
    function sizeOfInt64() public pure returns(uint8){return 8; }
}
```

**Comment by silasdavis** [comment]
@pouladzade I think this result of trying to read memory at a negative offset. Now if this is the case then it will be an error but we should avoid a panic (I will make a change to the memory to check this).

However, I believe the bytecode you are running probably runs elsewhere (remix), could you share the bytecode that causes the panic. A possible cause could be an unsupported opcode usage, or perhaps something else. I can get further if you provide the bytecode, but the fact this is a panic at all is noteworthy so thanks!

**Comment by pouladzade** [comment]
@silasdavis I tried to regenerate the bug but unfortunately I couldn't,  because on that time I was writing the solidity assembly and was testing my code, but now I completed the code and it works fine.
I could send the log only because the after calling my function, burrow got panic and never lunch anymore, I mean when I tried to run burrow it couldn't validate the chain's data anymore!

**Comment by pouladzade** [comment]
I wrote an assembly code which can generate this bug, You were right, it happens when the value of the loop index gets negative.
[StringTest.txt](https://github.com/hyperledger/burrow/files/1766696/StringTest.txt)
[compile-json.txt](https://github.com/hyperledger/burrow/files/1766698/compile-json.txt)
![bug-697](https://user-images.githubusercontent.com/10783040/36783214-4eb1ba38-1cb6-11e8-9839-c7a0ba1182bd.jpg)

**Comment by silasdavis** [comment]
There are two levels of solution to this:

1. We need to wrap `Execute(tx) error` with panic/recover so that any downstream panic in the execution engine is handled as an error. This addresses not only this issue but the entire class of 'panic on tx' issues - rather than crashing the chain we will just reject the transaction. I've been intending to do this for a while - but needed to fix our internal API a bit. Now is a good time to do it!

2. We need to (this is the royal we - I mean I need to) change our memory implementation so that it does not accept negative offsets: https://github.com/hyperledger/burrow/blob/develop/execution/execution.go#L153 - it probably makes sense to use `uint64` here, though they are not so good for arithmetic/have their own issues. In any case there will be some negative number checks.

**Comment by silasdavis** [comment]
fixed by #704
